### PR TITLE
[POAE7-2396]QueryDataGenerator using arrowArray/schema

### PR DESCRIPTION
**Pull request by imajiayu**
### What changes were proposed in this pull request?
Add QueryArrowDataGenerator.h, which is using ArrowArrayBuilder to generate arrow format of basic data types. 

### How was this patch tested?
QueryArrowDataGeneratorTest.cpp

### CODE REVIEW

File: cider/tests/utils/QueryArrowDataGeneratorTest.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    ArrowArray* expected_array = nullptr;                                           \
+    ArrowSchema* expected_schema = nullptr;                                         \
+    std::tie(expected_schema, expected_array) =                                     \
+        ArrowArrayBuilder()                                                         \
+            .addColumn<C_TYPE>("col_1", CREATE_SUBSTRAIT_TYPE(S_TYPE), VEC)         \
+            .build();                                                               \
+    EXPECT_EQ(expected_array->length, 10);                                          \
+    EXPECT_EQ(expected_array->n_children, 1);                                       \
+  }
+// EXPECT_TRUE(ArrowChecker::checkEq(expected_batch, actual_batch));
```
Comment (jikunshang) [inline cider/tests/utils/QueryArrowDataGeneratorTest.cpp:48]: remove or add some todo?

File: cider/tests/utils/QueryArrowDataGeneratorTest.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+          .addColumn<int16_t>("", CREATE_SUBSTRAIT_TYPE(I16), VEC)
+          .addColumn<int32_t>("", CREATE_SUBSTRAIT_TYPE(I32), VEC)
+          .addColumn<int64_t>("", CREATE_SUBSTRAIT_TYPE(I64), VEC)
+          .addColumn<float>("", CREATE_SUBSTRAIT_TYPE(Fp32), VEC)
+          .addColumn<double>("", CREATE_SUBSTRAIT_TYPE(Fp64), VEC)
+          .build();
+  // EXPECT_TRUE(ArrowChecker::checkEq(expected_batch, actual_batch));
+}
+
+int main(int argc, char** argv) {
```
Comment (jikunshang) [inline cider/tests/utils/QueryArrowDataGeneratorTest.cpp:129]: can you add some cases for null?

File: cider/tests/utils/QueryArrowDataGenerator.h [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                        // each row has 1/x chance to be null.
+      const T value_min = N_MIN,
+      const T value_max = N_MAX) {
+    std::vector<T> col_data(row_num);
+    std::vector<bool> null_data(row_num);
+    std::mt19937 rng(std::random_device{}());  // NOLINT
+    switch (pattern) {
+      case GeneratePattern::Sequence:
+        for (auto i = 0; i < row_num; ++i) {
+          null_data[i] = Random::oneIn(null_chance, rng) ? (col_data[i] = N_MIN, true)
```
Comment (jikunshang) [inline cider/tests/utils/QueryArrowDataGenerator.h:112]: we don't need this `col_data[i] = N_MIN` anymore. could use 0

Comment (imajiayu) [reply cider/tests/utils/QueryArrowDataGenerator.h:112]: do you mean replace "N_MIN" with "0" here?

Comment (jikunshang) [reply cider/tests/utils/QueryArrowDataGenerator.h]: yes `col_data[i] = 0` would be fine.

File: cider/tests/utils/QueryArrowDataGenerator.h [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+          GENERATE_AND_ADD_COLUMN(float)
+        case ::substrait::Type::KindCase::kFp64:
+          GENERATE_AND_ADD_COLUMN(double)
+        default:
+          CIDER_THROW(CiderCompileException, "Type not supported.");
+      }
+    }
+
+    auto schema_array = builder.build();
+    schema = std::get<0>(schema_array);
```
Comment (yma11) [inline cider/tests/utils/QueryArrowDataGenerator.h:92]: rename to `array_with_schema`?

File: cider/tests/utils/QueryArrowDataGenerator.h [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                                  false);
+          }
+        } else if (std::is_floating_point<T>::value) {
+          for (auto i = 0; i < col_data.size(); ++i) {
+            null_data[i] = Random::oneIn(null_chance, rng)
+                               ? (col_data[i] = N_MIN, true)
+                               : (col_data[i] = static_cast<T>(
+                                      Random::randFloat(value_min, value_max, rng)),
+                                  false);
+          }
```
Comment (yma11) [inline cider/tests/utils/QueryArrowDataGenerator.h:136]: Won't be okay to also add `double` here?

File: cider/tests/utils/QueryArrowDataGenerator.h (hunk 3 above)
Comment (yma11) [reply cider/tests/utils/QueryArrowDataGenerator.h]: We may need to verify a scenario that the value is actual ` N_MIN` but in null_vector, it's set to false, which means it's not a null value and should be selected. Should we add some logic here?

**Review by spevenhe — APPROVED** [review]
LGTM

**Review by kd97 — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]

**Pull request merged**